### PR TITLE
[MOB-82] Persist in-app messages on disk

### PR DESCRIPTION
**Pull request by vbabenkoru**

### CODE REVIEW

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableInAppFileStorage.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        JSONArray messagesJson = new JSONArray();
+
+        try {
+            for (Map.Entry<String, IterableInAppMessage> entry : messages.entrySet()) {
+                IterableInAppMessage message = entry.getValue();
+                messagesJson.put(message.toJSONObject());
+            }
+            jsonData.putOpt("inAppMessages", messagesJson);
+        } catch (JSONException ignored) {}
+
```
Comment (tapashmajumder) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableInAppFileStorage.java]: Log if exception?

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableInAppMessage.java [hunk 2]
```diff
@@ -55,30 +55,71 @@ boolean isConsumed() {
 
     void setConsumed(boolean consumed) {
         this.consumed = consumed;
+        onChanged();
     }
 
-    public static IterableInAppMessage fromJSON(IterableInAppStorage storage, JSONObject messageJson) {
+    static IterableInAppMessage fromJSONObject(JSONObject messageJson) {
         if (messageJson != null) {
```
Comment (tapashmajumder) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableInAppMessage.java]: Not needed for +1. Why not change the if clause to "if (messageJson == null) { return null;} " Then the main code body will be 1 level less indented.

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableInAppMessage.java [hunk 3]
```diff
@@ -55,30 +55,71 @@ boolean isConsumed() {
...
 
+    JSONObject toJSONObject() {
+        JSONObject messageJson = new JSONObject();
+        JSONObject contentJson = new JSONObject();
+        try {
+            messageJson.putOpt(IterableConstants.KEY_MESSAGE_ID, messageId);
+
+            contentJson.putOpt("html", content.html);
+            contentJson.putOpt("payload", content.payload);
+            contentJson.putOpt("inAppDisplaySettings", encodePaddingRectToJson(content.padding));
```
Comment (tapashmajumder) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableInAppMessage.java]: Not needed for +1. You may want to consider moving these to consts somewhere.

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableInAppMessage.java [hunk 4]
```diff
@@ -55,30 +55,71 @@ boolean isConsumed() {
...
+            if (content.backgroundAlpha != 0) {
+                contentJson.putOpt("backgroundAlpha", content.backgroundAlpha);
+            }
+
+            messageJson.putOpt(IterableConstants.ITERABLE_IN_APP_CONTENT, contentJson);
+            messageJson.putOpt("processed", processed);
+            messageJson.putOpt("consumed", consumed);
+        } catch (JSONException ignored) {}
+        return messageJson;
```
Comment (tapashmajumder) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableInAppMessage.java]: Log warning etc on exception?

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableUtil.java [hunk 5]
```diff
@@ -84,4 +93,42 @@ static JSONObject retrieveExpirableJsonObject(SharedPreferences preferences, Str
...
+            BufferedReader bufferedReader = new BufferedReader(streamReader);
+            StringBuilder stringBuilder = new StringBuilder();
+            String line;
+            while ((line = bufferedReader.readLine()) != null) {
+                stringBuilder.append(line);
+            }
+            return stringBuilder.toString();
+        } catch (Exception e) {
+            IterableLogger.e("IterableUtil", "Error while reading file: " + file.toString(), e);
+        }
```
[the remark this answers is not in the document]
Comment (vbabenkoru) [reply iterableapi/src/main/java/com/iterable/iterableapi/IterableUtil.java]: oops!

**Pull request merged**